File: 4-kyu/explosive-sum/explosive-sum.js
function sum(n){
  let p=[1];
  for(let i=1;i<=n;i++){
    let s=0;
    for(let k=1;;k++){
      let g1=k*(3*k-1)/2,g2=k*(3*k+1)/2;
      if(g1>i&&g2>i) break;
      let sign=k%2?1:-1;
      if(g1<=i) s+=sign*p[i-g1];
      if(g2<=i) s+=sign*p[i-g2];
    }
    p[i]=s;
  }
  return p[n];
}

File: 4-kyu/explosive-sum/explosive-sum.py
def exp_sum(n: int) -> int:
    if n < 0: return 0
    p = [1] + [0] * n
    for m in range(1, n + 1):
        k, total = 1, 0
        while True:
            g1, g2 = k*(3*k-1)//2, k*(3*k+1)//2
            if g1 > m: break
            sign = -1 if k % 2 == 0 else 1
            total += sign * p[m - g1]
            if g2 <= m: total += sign * p[m - g2]
            k += 1
        p[m] = total
    return p[n]